[KEP-4355] Coordinated Leader Election alpha documentation

Jefftree opened the pull request:
<!--

 Hello!

 Remember to ADD A DESCRIPTION and delete this note before submitting
 your pull request. The description should explain what will change,
 and why.

 PLEASE title the FIRST commit appropriately, so that if you squash all
 your commits into one, the combined commit message makes sense.
 For overall help on editing and submitting pull requests, visit:
  https://kubernetes.io/docs/contribute/suggesting-improvements/

 Use the default base branch, “main”, if you're documenting existing
 features in the English localization.

 If you're working on a different localization (not English), see
 https://kubernetes.io/docs/contribute/new-content/overview/#choose-which-git-branch-to-use
 for advice.

 If you're documenting a feature that will be part of a future release, see
 https://kubernetes.io/docs/contribute/new-content/new-features/ for advice.

-->
https://github.com/kubernetes/enhancements/issues/4355#issuecomment-2181407075

chanieljdan commented:
Hello @Jefftree 👋, 1.31 Docs Shadow here,

A quick reminder to get your PR ready for review before next week, Tuesday July 16th 2024 18:00 PST. For additional information, refer to: [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review).

Thank you!

Daniel

Princesso commented:
Hi @Jefftree, a gentle reminder that tomorrow is the deadline for having your Docs PR ready for review. Please take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before tomorrow, Tuesday, July 16th, 2024 18:00 PST.

chanieljdan commented:
Hi @Jefftree, a gentle reminder that today is the deadline for having your Docs PR ready for review. Please take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before today at 18:00 PST.

Princesso commented:
Hi @Jefftree can you move this PR out of drafts? That way, more reviewers can start looking at it.

Princesso commented:
Hello @Jefftree 

v1.31 Doc Lead here. This PR has been marked as at risk for Doc Freeze.

Please note that the Doc freeze deadline is tomorrow, and if the PR is not merged by tomorrow, you will need to file an exception or this enhancement will be removed from this release. Thank you!

Jefftree commented:
/assign @sttts @aojea

sttts commented on content/en/docs/concepts/architecture/coordinated-leader-election.md:
```diff
@@ -0,0 +1,56 @@
+
+Kubernetes {{< skew currentVersion >}} includes an alpha feature that allows
+components to deterministically select a leader via Coordinated Leader Election.
+This is useful to satisfy Kubernetes version skew during cluster upgrades.
```
```suggestion
This is useful to satisfy Kubernetes version skew constraints during cluster upgrades.
```

sttts commented on content/en/docs/concepts/architecture/coordinated-leader-election.md:
```diff
@@ -0,0 +1,56 @@
+Kubernetes {{< skew currentVersion >}} includes an alpha feature that allows
+components to deterministically select a leader via Coordinated Leader Election.
+This is useful to satisfy Kubernetes version skew during cluster upgrades.
+Currently, the only supported selection strategy is `OldestEmulationVersion`,
```
```suggestion
Currently, the only supported builtin selection strategy is `OldestEmulationVersion`,
```

sttts commented on content/en/docs/concepts/architecture/coordinated-leader-election.md:
```diff
@@ -0,0 +1,56 @@
+enabled.
+
+This can be done by setting `FEATURE_GATES="CoordinatedLeaderElection=true"` and
+`RUNTIME_CONFIG="coordination.k8s.io/v1alpha1=true"`.
```
in which context do we support env vars for that? The general interface are command line flags.

Jefftree replied:
Updated to use flags instead of env vars.

sttts commented on content/en/docs/concepts/architecture/coordinated-leader-election.md:
```diff
@@ -0,0 +1,56 @@
+and Lease goroutine (both found in client-go/pkg/leaderelection). Two components
+(kube-controller-manager and kube-scheduler) will automatically use coordinated
+leader election if enabled. Please refer to the example found in
+`k8s.io/cmd/kube-scheduler/app/server.go` on set up.
```
```suggestion
`k8s.io/kubernetes/cmd/kube-scheduler/app/server.go` on set up.
```

How about a link instead?

sttts commented on content/en/docs/concepts/architecture/coordinated-leader-election.md:
````diff
@@ -0,0 +1,56 @@
+  renewTime: "2024-07-30T03:45:18.325483Z"
+```
+
+Please refer to the documentation for LeaseCandidate for the full API details.
````
link?

Jefftree replied:
added

Princesso commented:
Hello @Jefftree

v1.31 Doc Lead here. This PR is still at risk for Doc Freeze because it is yet to be merged.
Please implement the changes required by the reviewers to have another round of reviews.

Please note that if the PR is not merged by 18:00 PDT today, you will need to file an exception for the associated enhancement to be a part of this milestone. Thank you!

Jefftree reviewed (commented):
Fixed comments

sttts commented:
Lgtm from technical point of view.

Princesso commented:
/assign @sftim

chanieljdan commented:
Technical LGTM provided [above](https://github.com/kubernetes/website/pull/46982#issuecomment-2258640710)
/lgtm

sftim commented on content/en/docs/concepts/architecture/coordinated-leader-election.md:
```diff
@@ -0,0 +1,56 @@
+preferring the leader with the lowest emulation version, followed by binary
+version, followed by creation timestamp.
+
+## Enabling Coordinated Leader Election
```
(nit, OK for alpha, please fix for beta)
```suggestion
## Enabling coordinated leader election
```

sftim commented on content/en/docs/concepts/architecture/coordinated-leader-election.md:
```diff
@@ -0,0 +1,56 @@
+This can be done by setting flags `--feature-gates="CoordinatedLeaderElection=true"` and
+`--runtime-config="coordination.k8s.io/v1alpha1=true"`.
+
+## Component Configuration
```
(nit, OK for alpha, please fix for beta)
```suggestion
## Component configuration
```

sftim commented on content/en/docs/concepts/architecture/coordinated-leader-election.md:
````diff
@@ -0,0 +1,56 @@
+
+## Component Configuration
+
+With Coordinated Leader Election, components need to both run a LeaseCandidate
+and Lease goroutine (both found in client-go/pkg/leaderelection). Two components
+(kube-controller-manager and kube-scheduler) will automatically use coordinated
+leader election if enabled. Please refer to the example found in
+[k8s.io/kubernetes/cmd/kube-scheduler/app/server.go](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-scheduler/app/server.go) on set up.
+
+The created LeaseCandidate object looks similar to below:
+
+```
+apiVersion: coordination.k8s.io/v1alpha1
+kind: LeaseCandidate
+metadata:
+  name: hostname_uuid
+  namespace: kube-system
+spec:
+  binaryVersion: 1.31.0
+  emulationVersion: 1.31.0
+  leaseName: kube-scheduler
+  preferredStrategies:
+  - OldestEmulationVersion
+  renewTime: "2024-07-30T03:45:18.325483Z"
+```
+
+Please refer to the [documentation](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#leasecandidate-v1alpha1-coordination-k8s-io) for LeaseCandidate for the full API details.
````
If I'm running a cluster, do I need to read that source code? If not, try:
```suggestion
Provided that you have enabled the `CoordinatedLeaderElection` feature gate _and_ have
the `coordination.k8s.io/v1alpha1` API group enabled, control plane components automatically
use the {{< api-reference page="cluster-resources/lease-v1" >}} and
{{< api-reference page="cluster-resources/lease-candidate-v1alpha1" >}} APIs to elect a leader
as needed.
```

This is a problem though as we don't yet have even a placeholder API reference page for LeaseCandidate.
I suggest adding one and making it clear in this PR that the page is a placeholder to be overwritten when we generate a new API reference.

See https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/

Jefftree replied:
added

sftim commented on content/en/docs/reference/command-line-tools-reference/feature-gates/coordinated-leader-election.md:
```diff
@@ -0,0 +1,14 @@
+    defaultValue: false
+    fromVersion: "1.31"
+---
+Enables coordinated leader election for leases, deterministically
```
```suggestion
Enables the behaviors supporting the LeaseCandidate API, and also enables coordinated leader election
for the Kubernetes control plane, deterministically
```

sftim commented on content/en/docs/concepts/architecture/coordinated-leader-election.md:
```diff
@@ -0,0 +1,56 @@
+{{< feature-state feature_gate_name="CoordinatedLeaderElection" >}}
+
+Kubernetes {{< skew currentVersion >}} includes an alpha feature that allows
+components to deterministically select a leader via Coordinated Leader Election.
```
```suggestion
{{< glossary_tooltip text="control plane" term_id="control-plane" >}} components to
deterministically select a leader via _coordinated leader election_.
```
This doesn't look useful for clients using Lease that aren't part of the control plane, but maybe I'm wrong here?

sftim commented on content/en/docs/concepts/architecture/coordinated-leader-election.md:
I would put this page inside /docs/concepts/cluster-administration/ as it is a detail that I think only cluster administrators would need to know about.

Jefftree replied:
updated

sftim reviewed (commented):
Thanks

Please also edit https://kubernetes.io/docs/concepts/architecture/leases/ to link to the new page.

sftim commented on content/en/docs/concepts/cluster-administration/coordinated-leader-election.md:
```diff
@@ -0,0 +1,56 @@
+{{< feature-state feature_gate_name="CoordinatedLeaderElection" >}}
+
+Kubernetes {{< skew currentVersion >}} includes an alpha feature that allows
+{{< glossary_tooltip text="control plane" term_id="control-plane" >}} components to deterministically select a leader via _coordinated leader election.
```
Oops, I might've missed a character in a previous suggestion
```suggestion
{{< glossary_tooltip text="control plane" term_id="control-plane" >}} components to
deterministically select a leader via _coordinated leader election_.
```

sftim commented on content/en/docs/concepts/cluster-administration/coordinated-leader-election.md:
```diff
@@ -0,0 +1,56 @@
+This can be done by setting flags `--feature-gates="CoordinatedLeaderElection=true"` and
+`--runtime-config="coordination.k8s.io/v1alpha1=true"`.
+
+## Component Configuration
```
(nit)
```suggestion
## Component configuration
```

sftim commented on content/en/docs/concepts/cluster-administration/coordinated-leader-election.md:
```diff
@@ -0,0 +1,56 @@
+Ensure that `CoordinatedLeaderElection` [feature
+gate](/docs/reference/command-line-tools-reference/feature-gates/) is enabled
+when you start the {{< glossary_tooltip text="API Server"
+term_id="kube-apiserver" >}}: and that the `coordination.k8s.io/v1alpha1` API is
```
(nits)
```suggestion
term_id="kube-apiserver" >}}, and that the `coordination.k8s.io/v1alpha1` API group is
```

sftim commented on content/en/docs/concepts/cluster-administration/coordinated-leader-election.md:
````diff
@@ -0,0 +1,56 @@
+
+## Component Configuration
+
+With Coordinated Leader Election, components need to both run a LeaseCandidate
+and Lease goroutine (both found in client-go/pkg/leaderelection). Two components
+(kube-controller-manager and kube-scheduler) will automatically use coordinated
+leader election if enabled. Please refer to the example found in
+[k8s.io/kubernetes/cmd/kube-scheduler/app/server.go](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-scheduler/app/server.go) on set up.
+
+The created LeaseCandidate object looks similar to below:
+
+```
+apiVersion: coordination.k8s.io/v1alpha1
+kind: LeaseCandidate
+metadata:
+  name: hostname_uuid
+  namespace: kube-system
+spec:
+  binaryVersion: 1.31.0
+  emulationVersion: 1.31.0
+  leaseName: kube-scheduler
+  preferredStrategies:
+  - OldestEmulationVersion
+  renewTime: "2024-07-30T03:45:18.325483Z"
+```
````
```suggestion
Provided that you have enabled the `CoordinatedLeaderElection` feature gate _and_
have the `coordination.k8s.io/v1alpha1` API group enabled, compatible control plane
components automatically use the LeaseCandidate and Lease APIs to elect a leader
as needed.

For Kubernetes {{< skew currentVersion >}}, two control plane components
(kube-controller-manager and kube-scheduler) automatically use coordinated
leader election when the feature gate and API group are enabled.
```

(see https://github.com/kubernetes/website/pull/46982#discussion_r1697286771 - we shouldn't be directing users to read source code).

---

If you'd like to write about this on our [contributor blog](https://k8s.dev/blog/), please do - it's a nice feature, but the whole idea is to make life simpler for end users so they don't need to care about how it works.

Jefftree replied:
ack, definitely interested in doing a write up around the more technical details without having it all littered in the documentation for end users.

sftim replied:
https://github.com/kubernetes/contributor-site is the place to submit the blog PR, BTW

sftim commented on content/en/docs/reference/kubernetes-api/cluster-resources/lease-candidate-v1alpha1.md:
```diff
@@ -0,0 +1,12 @@
+  kind: "LeaseCandidate"
+content_type: "api_reference"
+description: "This is a placeholder page to be overwritten when the docs are generated for 1.31"
+title: "Lease Candidate"
```
```suggestion
title: "LeaseCandidate"
```
(may look weird otherwise)

Jefftree commented:
@sftim ah I'm so sorry I renamed the file and addressed half your comments in the deleted file so they were marked as resolved without being fixed. 🤦 Should be fixed now.

sftim reviewed (commented):
Thanks for getting this ready.

/lgtm
/approve
